### PR TITLE
Add message meta data support

### PR DESCRIPTION
**Pull request by msiebeneicher**
This PR is related to #82 

- [x] Add a "key" to the Message struct ([]byte)
- [x] Consumers reading from key/value stores should set the key
- [x] Producers requiring a key should take the key field into account (can be empty)
- [ ] Modulators for keys must be possible (ideally by defining an ApplyTo field or sth. similar)
- [ ] Implement a "Copy" formatter that allows copying data between key and value fields

### CODE REVIEW

File: producer/kafka.go [hunk 1]
```diff
@@ -507,15 +500,12 @@ func (prod *Kafka) produceMessage(msg *core.Message) {
...
+		Metadata: &msg,
 	}
 
-	if prod.keyFirst {
-		keyMsg := originalMsg.Clone()
-		kafkaMsg.Key = kafka.ByteEncoder(keyMsg.Data())
-	} else {
-		keyMsg := msg.Clone()
-		kafkaMsg.Key = kafka.ByteEncoder(keyMsg.Data())
+	kafkaKey := msg.MetaData().GetValue("key", []byte(""))
```
Comment (arnecls) [inline producer/kafka.go]: make "key" a config option (defaults to "key")

File: consumer/kafka.go [hunk 2]
```diff
@@ -215,6 +216,10 @@ func (cons *Kafka) Configure(conf core.PluginConfigReader) error {
 	cons.offsets = make(map[int32]*int64)
 	cons.MaxPartitionID = 0
 	cons.keySeparator = []byte(conf.GetString("KeySeparator", ":"))
+
+	if conf.HasValue("PrependKey") {
```
Comment (arnecls) [inline consumer/kafka.go]: Remove it. We're 0.x.y for a reason ;)

File: core/messagehelper.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+// DropMessage restore the original message and "drops" they the active message stream router
+func DropMessage(msg *Message) error {
+	return DropMessageByRouter(msg, msg.GetRouter())
+}
+
+// DropMessageByRouter restore the original message and "drops" they to specific router
+func DropMessageByRouter(msg *Message, router Router) error {
+	CountDroppedMessage()
+	return Route(msg.CloneOriginal(), router)
```
Comment (arnecls) [inline core/messagehelper.go]: error handling should move here.
log.Error and continue

File: core/message.go [hunk 4]
```diff
@@ -203,6 +241,11 @@ func (msg *Message) Cap() int {
 	return cap(msg.data.payload)
 }
 
+// MetaData returns the current MetaData
+func (msg *Message) MetaData() *MetaData {
```
Comment (arnecls) [inline core/message.go]: Metadata already is a reference type.

File: core/messagehelper.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	"strings"
+)
+
+// GetAppliedContent is a func() to get message content from payload or meta data
+// for later handling by plugins
+type GetAppliedContent func(msg *Message) []byte
+
+// GetAppliedContentFunction returns a GetAppliedContent function
+func GetAppliedContentFunction(conf PluginConfigReader) GetAppliedContent {
+	parts := strings.Split(conf.GetString("ApplyTo", ""), ":")
```
Comment (arnecls) [inline core/messagehelper.go]: pass in the string directly?
Removes the dependency to a config

File: core/messagehelper.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+	if parts[0] == "meta" {
+		return func(msg *Message) []byte {
+			return msg.MetaData().GetValue(parts[1], []byte(""))
```
Comment (arnecls) [inline core/messagehelper.go]: []byte{}

**Comment by msiebeneicher** [comment]
Requested changes are done. How discussed i will merge this PR and create two new issues to `update the modulators for meta data usage` and to `create a new formatter for setting meta data`.

**Pull request merged**